core/src/test/java/org/apache/brooklyn/core/config/ConfigKeyDeprecationRebindTest.java

```diff
@@ -0,0 +1,373 @@
+
+    @Test
+    public void testUsingDeprecatedName() throws Exception {
+        MyApp entity = app().addChild(EntitySpec.create(MyApp.class)
+                .configure("oldKey1", "myval"));
+        assertEquals(entity.config().get(MyApp.KEY_1), "myval");
+        
+        rebind();
+        
+        Entity newEntity = mgmt().getEntityManager().getEntity(entity.getId());
+        assertEquals(newEntity.config().get(MyApp.KEY_1), "myval");
+        
+        // Expect the persisted state of the entity to have used the non-deprecated name.
+        String allLines = getPersistanceFileContents(BrooklynObjectType.ENTITY, newEntity.getId());
+        assertFalse(allLines.contains("oldKey1"), "contains 'oldKey1', allLines="+allLines);
+        assertTrue(allLines.contains("key1"), "contains 'key1', allLines="+allLines);
+    }
+    
+    /**
+     * Created with:
+     * <pre>
+     * {@code
+     * Entity app = mgmt().getEntityManager().createEntity(EntitySpec.create(MyApp.class)
+     *         .configure("oldKey1", "myval"));
+     * }
+     * </pre>
+     */
+    @Test
+    public void testEntityPersistedWithDeprecatedKeyName() throws Exception {
+        String appId = "pps2ttgijb";
+        addMemento(BrooklynObjectType.ENTITY, "config-deprecated-key", appId);
+        rebind();
+        
+        Entity newApp = mgmt().getEntityManager().getEntity(appId);
+        assertEquals(newApp.config().get(MyApp.KEY_1), "myval");
+        
+        // Expect the persisted state to have been re-written with the new key value.
+        switchOriginalToNewManagementContext();
+        rebind();
+        
+        String allLines = getPersistanceFileContents(BrooklynObjectType.ENTITY, appId);
+        assertFalse(allLines.contains("oldKey1"), "should not contain 'oldKey1', allLines="+allLines);
+        assertTrue(allLines.contains("<key1>"), "should contain '<key1>', allLines="+allLines);
+    }
+
+    /**
+     * Created with:
+     * <pre>
+     * {@code
+     * Entity app = mgmt().getEntityManager().createEntity(EntitySpec.create(MyApp.class)
+     *         .configure("oldFlagKey2", "myval"));
+     * }
+     * </pre>
+     */
+    @Test
+    public void testEntityPersistedWithSetFromFlagNameOnKey() throws Exception {
+        String appId = "ug77ek2tkd";
+        addMemento(BrooklynObjectType.ENTITY, "config-deprecated-flagNameOnKey", appId);
+        rebind();
+        
+        Entity newApp = mgmt().getEntityManager().getEntity(appId);
+        assertEquals(newApp.config().get(MyApp.KEY_2), "myval");
+        
+        // Expect the persisted state to have been re-written with the new key value.
+        switchOriginalToNewManagementContext();
+        rebind();
+        
+        String allLines = getPersistanceFileContents(BrooklynObjectType.ENTITY, appId);
+        assertFalse(allLines.contains("oldFlagKey2"), "should not contain 'oldFlagKey2', allLines="+allLines);
```

Comment:
Yeah, it's a bit of a weird test. It was sensible to decide to write it - I generated the persisted state (as per the javadoc) with:
```
Entity app = mgmt().getEntityManager().createEntity(EntitySpec.create(MyApp.class)
        .configure("oldFlagKey2", "myval"));
```

That turns out not to contain `oldFlagKey2` because it was transformed at the time of writing the persisted state (even before any changes in this PR).

I think it's important that we record that this upgrade-path works. I could delete the test, and replace it with just a comment in the test class to say why we're not explicitly testing this scenario.

Or I could leave in the test, and add to the javadoc an explanation of why it's weird.

Thoughts @geomacy?